File: go/indicators/common/rateofchangepercent/rateofchangepercent.go
package rateofchangepercent

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// RateOfChangePercent is the difference between today's sample and the sample ℓ periods ago
// scaled by the old sample so as to represent the increase as a fraction.
//
// The values are centered at zero and can be positive and negative.
//
// ROC%ᵢ = (Pᵢ - Pᵢ₋ℓ) / Pᵢ₋ℓ = (Pᵢ/Pᵢ₋ℓ -1),
//
// where ℓ is the length.
//
// The indicator is not primed during the first ℓ updates.
type RateOfChangePercent struct {
	mu sync.RWMutex
	core.LineIndicator
	window       []float64
	windowLength int
	windowCount  int
	lastIndex    int
	primed       bool
}

// New returns an instance of the indicator created using supplied parameters.
func New(p *Params) (*RateOfChangePercent, error) {
	const (
		invalid = "invalid rate of change percent parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtn    = "rocp(%d%s)"
		minlen  = 1
	)

	length := p.Length
	if length < minlen {
		return nil, fmt.Errorf(fmts, invalid, "length should be positive")
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	// Resolve defaults for component functions.
	// A zero value means "use default, don't show in mnemonic".
	bc := p.BarComponent
	qc := p.QuoteComponent
	tc := p.TradeComponent

	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf(fmtn, length, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Rate of Change Percent " + mnemonic

	r := &RateOfChangePercent{
		window:       make([]float64, length+1),
		windowLength: length + 1,
		lastIndex:    length,
	}

	r.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, r.Update)

	return r, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *RateOfChangePercent) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
// It always has a single scalar output -- the calculated value of the rate of change percent.
func (s *RateOfChangePercent) Metadata() core.Metadata {
	return core.Metadata{
		Type:        core.RateOfChangePercent,
		Mnemonic:    s.LineIndicator.Mnemonic,
		Description: s.LineIndicator.Description,
		Outputs: []outputs.Metadata{
			{
				Kind:        int(Value),
				Type:        outputs.ScalarType,
				Mnemonic:    s.LineIndicator.Mnemonic,
				Description: s.LineIndicator.Description,
			},
		},
	}
}

// Update updates the value of the indicator given the next sample.
//
// The indicator is not primed during the first ℓ updates.
func (s *RateOfChangePercent) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	const epsilon = 1e-13

	if s.primed {
		if s.lastIndex > 1 {
			for i := 0; i < s.lastIndex; i++ {
				s.window[i] = s.window[i+1]
			}
		}

		s.window[s.lastIndex] = sample
		previous := s.window[0]
		if math.Abs(previous) > epsilon {
			return sample/previous - 1
		}

		return 0
	}

	s.window[s.windowCount] = sample
	s.windowCount++

	if s.windowLength == s.windowCount {
		s.primed = true
		previous := s.window[0]
		if math.Abs(previous) > epsilon {
			return sample/previous - 1
		}

		return 0
	}

	return math.NaN()
}

File: ts/indicators/common/rate-of-change-percent/rate-of-change-percent.ts
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { RateOfChangePercentOutput } from './rate-of-change-percent-output';
import { RateOfChangePercentParams } from './rate-of-change-percent-params';

/**
 * __Rate of Change Percent__ (__ROCP__) is the difference between today's sample and the sample ℓ periods ago
 * scaled by the old sample so as to represent the increase as a fraction.
 *
 * The values are centered at zero and can be positive and negative.
 *
 *     ROC%ᵢ = (Pᵢ - Pᵢ₋ℓ) / Pᵢ₋ℓ = (Pᵢ/Pᵢ₋ℓ - 1),
 *
 * where ℓ is the length.
 *
 * The indicator is not primed during the first ℓ updates.
 */
export class RateOfChangePercent extends LineIndicator {
  private readonly window: number[];
  private readonly windowLength: number;
  private readonly lastIndex: number;
  private windowCount = 0;

  /** Constructs an instance given the parameters. */
  public constructor(params: RateOfChangePercentParams) {
    super();
    const length = Math.floor(params.length);
    if (length < 1) {
      throw new Error('length should be positive');
    }

    this.window = new Array<number>(length + 1).fill(0);
    this.windowLength = length + 1;
    this.lastIndex = length;

    const mn = 'rocp(' + length.toString() +
      componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent) + ')';

    this.mnemonic = mn;
    this.description = 'Rate of Change Percent ' + mn;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;
    this.primed = false;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.RateOfChangePercent,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: RateOfChangePercentOutput.RateOfChangePercentValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the value of the rate of change percent given the next sample. */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return sample;
    }

    const epsilon = 1e-13;

    if (this.primed) {
      if (this.lastIndex > 1) {
        for (let i = 0; i < this.lastIndex; i++) {
          this.window[i] = this.window[i + 1];
        }
      }

      this.window[this.lastIndex] = sample;
      const previous = this.window[0];
      if (Math.abs(previous) > epsilon) {
        return sample / previous - 1;
      }

      return 0;
    }

    this.window[this.windowCount] = sample;
    this.windowCount++;

    if (this.windowLength === this.windowCount) {
      this.primed = true;
      const previous = this.window[0];
      if (Math.abs(previous) > epsilon) {
        return sample / previous - 1;
      }

      return 0;
    }

    return Number.NaN;
  }
}
